### DISCUSSION TITLE
Missing filePicker option for IPadOs?

### DISCUSSION THREAD

**Discussion by Sharkboy-j** [AvaloniaUI/Avalonia#8752] [Q&A] [answered]
Well, Im trying to open file dialog using 

`var dialog = new OpenFileDialog();`

And it works on **Desktop** applications. Also it says that class is Obsolete. Use Windows.StorageProvider
According to this [post](https://github.com/AvaloniaUI/Avalonia/pull/8303) i have to write smth like this:
`var folder = await window.StorageProvider.OpenFolderPickerAsync(new FolderPickerOpenOptions());`

But here is question. Where do i find `window` object? Application starts on **IpadOs** as **ISingleViewApplicationLifetime** and it doesn't have any windows... I had try to create new instance of `window` and call `StorageProvider.OpenFolderPickerAsync()` but as expected, it throwing `NotSupportedOSException`

**Comment by maxkatz6** [comment, marked as answer]
It requires the TopLevel, which can be a window or a single view top level control. 
var topLevel = ((TopLevel)view.VisualRoot);

**Reply by Sharkboy-j** [reply]
> It requires the TopLevel, which can be a window or a single view top level control.
> var topLevel = ((TopLevel)view.VisualRoot);

Cool, thanks! But 'VisualRoot' marked as protected. Is there any way i can access it from viewmodel ?

**Reply by maxkatz6** [reply]
Try GetVisualRoot()

**Reply by maxkatz6** [reply]
It works for non desktop and very much possible. 

If GetVisualRoot returns null, then this control wasn't yet completely initialized. You can wait for the OnAttachesToVisualTree event in this case.

> I'm trying to do that during the OnFrameworkInitializationCompleted() method in the App class.

That's a problem. At this point framework is initialized, but toplevel/window is not. It's not yet opened/attached to the native host. 
Why would you even need to open a file picker from this method? Right on the startup?

> singleViewPlatform.MainView as TopLevel

MainView is a main view, it can be embedded in the top level, but not necessarily the top level.

Also while ApplicationLifetime.MainView/MainWindow can be used to show the picker for both desktop and mobile,  I wouldn't recommend this. Simply because your file picker might need to be opened from another window/view, but this code will always open it from the main one. Blocking it and focusing. 
Instead I recommend to rely on specific page/control from which you open the dialog. Like `button.GetVisualRoot() as TopLevel`

**Reply by maxkatz6** [reply]
If you have more questions, please open another discussion. This one is already answered.
